File: src/main/java/me/kikugie/techutils/event/KeyCallbacks.java
package me.kikugie.techutils.event;

import fi.dy.masa.litematica.data.DataManager;
import fi.dy.masa.malilib.hotkeys.IHotkeyCallback;
import fi.dy.masa.malilib.hotkeys.IKeybind;
import fi.dy.masa.malilib.hotkeys.KeyAction;
import me.kikugie.techutils.config.Configs;
import me.kikugie.techutils.config.InGameNotifier;
import me.kikugie.techutils.render.gui.ConfigGui;
import net.minecraft.client.MinecraftClient;
import net.minecraft.util.BlockMirror;
import net.minecraft.util.BlockRotation;


public class KeyCallbacks {
    public static void init() {
        Configs.LitematicConfigs.ROTATE_PLACEMENT.getKeybind().setCallback(new RotatePlacementCallback());
        Configs.LitematicConfigs.MIRROR_PLACEMENT.getKeybind().setCallback(new MirrorPlacementCallback());
        Configs.MiscConfigs.OPEN_CONFIG.getKeybind().setCallback(new OpenConfig());
    }

    private static class OpenConfig implements IHotkeyCallback {

        @Override
        public boolean onKeyAction(KeyAction action, IKeybind key) {
            if (key != Configs.MiscConfigs.OPEN_CONFIG.getKeybind())
                return false;
            MinecraftClient.getInstance().setScreen(new ConfigGui());
            return true;
        }
    }

    private static class RotatePlacementCallback implements IHotkeyCallback {
        @Override
        public boolean onKeyAction(KeyAction action, IKeybind key) {
            if (key != Configs.LitematicConfigs.ROTATE_PLACEMENT.getKeybind())
                return false;

            var placement = DataManager.getSchematicPlacementManager().getSelectedSchematicPlacement();
            if (placement == null)
                return false;
            placement.setRotation(placement.getRotation().rotate(BlockRotation.CLOCKWISE_90), InGameNotifier.INSTANCE);
            return true;
        }
    }

    private static class MirrorPlacementCallback implements IHotkeyCallback {
        @Override
        public boolean onKeyAction(KeyAction action, IKeybind key) {
            if (key != Configs.LitematicConfigs.MIRROR_PLACEMENT.getKeybind()) return false;

            var placement = DataManager.getSchematicPlacementManager().getSelectedSchematicPlacement();
            if (placement == null) return false;
            BlockMirror mirror = switch (placement.getMirror()) {
                case NONE -> BlockMirror.LEFT_RIGHT;
                case LEFT_RIGHT -> BlockMirror.FRONT_BACK;
                case FRONT_BACK -> BlockMirror.NONE;
            };
            placement.setMirror(mirror, InGameNotifier.INSTANCE);
            return true;
        }
    }
}

File: src/main/kotlin/me/kikugie/techutils/event/KeyCallbacks.kt
package me.kikugie.techutils.event

import fi.dy.masa.litematica.data.DataManager
import fi.dy.masa.litematica.schematic.placement.SchematicPlacement
import fi.dy.masa.malilib.hotkeys.IHotkeyCallback
import fi.dy.masa.malilib.hotkeys.IKeybind
import fi.dy.masa.malilib.hotkeys.KeyAction
import me.kikugie.techutils.config.Configs
import me.kikugie.techutils.config.InGameNotifier
import net.minecraft.util.BlockMirror
import net.minecraft.util.BlockRotation

object KeyCallbacks {
    fun init() {
        Configs.LitematicConfigs.ROTATE_PLACEMENT.keybind.setCallback(RotatePlacementCallback())
        Configs.LitematicConfigs.MIRROR_PLACEMENT.keybind.setCallback(MirrorPlacementCallback())
    }

    private class RotatePlacementCallback : IHotkeyCallback {
        override fun onKeyAction(action: KeyAction, key: IKeybind): Boolean {
            if (key === Configs.LitematicConfigs.ROTATE_PLACEMENT.keybind) {
                val placement: SchematicPlacement =
                    DataManager.getSchematicPlacementManager().selectedSchematicPlacement
                        ?: return false
                placement.setRotation(
                    placement.rotation.rotate(BlockRotation.CLOCKWISE_90),
                    InGameNotifier.INSTANCE
                )
                return true
            }
            return false
        }
    }

    private class MirrorPlacementCallback : IHotkeyCallback {
        override fun onKeyAction(action: KeyAction, key: IKeybind): Boolean {
            if (key === Configs.LitematicConfigs.MIRROR_PLACEMENT.keybind) {
                val placement: SchematicPlacement =
                    DataManager.getSchematicPlacementManager().selectedSchematicPlacement
                        ?: return false

                @Suppress("WHEN_ENUM_CAN_BE_NULL_IN_JAVA")
                val mirror: BlockMirror = when (placement.mirror) {
                    BlockMirror.NONE -> BlockMirror.LEFT_RIGHT
                    BlockMirror.LEFT_RIGHT -> BlockMirror.FRONT_BACK
                    BlockMirror.FRONT_BACK -> BlockMirror.NONE
                }
                placement.setMirror(mirror, InGameNotifier.INSTANCE)
                return true
            }
            return false
        }
    }
}